### PR TITLE
PayPal Web Integration

### PR DESCRIPTION
**Pull request by sarahkoop**
### Summary of changes

 - Remove Native Checkout SDK and implement a web flow for PayPal
 - Use browser switch for web flow
 - Update PayPalClient integration to use a listener pattern for results
 - Add PayPalError class and re-organize errors
 - General code/doc string cleanup throughout 

 ### Checklist

 - ~[ ] Added a changelog entry~

- @sarahkoop 
- @jcnoriega 
- @andrespelaezp

### CODE REVIEW

File: PayPalCheckout/src/main/java/com/paypal/android/checkout/BrowserSwitchHelper.kt [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import org.json.JSONObject
+
+
+internal class BrowserSwitchHelper(private val urlScheme: String) {
+
+    private val redirectUriPayPalCheckout = String.format("%s://x-callback-url/paypal-sdk/paypal-checkout", urlScheme)
+
+    private fun buildPayPalCheckoutUri(@NonNull orderId: String?, @NonNull config: CoreConfig): Uri {
+        val baseURL = when (config.environment) {
+            Environment.LIVE -> "https://www.paypal.com"
```
Comment (sarahkoop) [inline PayPalCheckout/src/main/java/com/paypal/android/checkout/BrowserSwitchHelper.kt:16]: In iOS we pulled these URLs into an Environment extension in the PayPal module - this doesn't exist in Kotlin, do we think these are ok to live in this helper class or should they go elsewhere?

File: PayPalCheckout/src/main/java/com/paypal/android/checkout/PayPalCheckoutListener.java [hunk 2]
```diff
@@ -2,6 +2,9 @@
 
 import androidx.annotation.NonNull;
 
-public interface PayPalCheckoutResultCallback {
+/**
+ * Interface for receiving results of the PayPal payment flow.
+ */
+public interface PayPalCheckoutListener {
```
Comment (sarahkoop) [inline PayPalCheckout/src/main/java/com/paypal/android/checkout/PayPalCheckoutListener.java]: We have some mis-aligned naming throughout this PR. We need to get consensus on what we want to call classes in this module, and the whole module (PayPal, PayPalCheckout ...)

File: PayPalCheckout/src/main/java/com/paypal/android/checkout/PayPalClient.kt [hunk 3]
```diff
@@ -1,44 +1,96 @@
...
+        configuration: CoreConfig,
+        urlScheme: String
+    ) : this(activity, configuration, BrowserSwitchClient(), BrowserSwitchHelper(urlScheme))
+
+    private var browserSwitchResult: BrowserSwitchResult? = null
+    var listener: PayPalCheckoutListener? = null
+        set(value) {
+            field = value
+            browserSwitchResult?.also {
+                handleBrowserSwitchResult()
```
Comment (sarahkoop) [inline PayPalCheckout/src/main/java/com/paypal/android/checkout/PayPalClient.kt:43]: This handles a browser switch immediately when the listener is set - this allows merchants to set a listener after their Activity has been resumed, and the browser switch has been delivered, and still receive the result. How should should this behavior be documented?

File: PayPalCheckout/src/main/java/com/paypal/android/checkout/PayPalClient.kt [hunk 4]
```diff
@@ -1,44 +1,100 @@
...
+    private var browserSwitchResult: BrowserSwitchResult? = null
+
+    var listener: PayPalCheckoutListener? = null
+        /**
+         * @param value a [PayPalCheckoutListener] to receive results from the PayPal flow
+         */
+        set(value) {
+            field = value
+            browserSwitchResult?.also {
+                handleBrowserSwitchResult()
```
Comment (sarahkoop) [inline PayPalCheckout/src/main/java/com/paypal/android/checkout/PayPalClient.kt:45]: This handles a browser switch immediately when the listener is set - this allows merchants to set a listener after their Activity has been resumed, and the browser switch has been delivered, and still receive the result. How should should this behavior be documented?

File: PayPalCheckout/src/main/java/com/paypal/android/checkout/error/PayPalError.kt [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    )
+
+    // 1. Result did not contain the expected data.
+    val malformedResultError = CoreSDKError(
+        code = Code.MALFORMED_RESULT.ordinal,
+        errorDescription = "Result did not contain the expected data. Payer ID or Order ID is null."
+    )
+}
+
+internal enum class Code {
```
Comment (sarahkoop) [inline PayPalCheckout/src/main/java/com/paypal/android/checkout/error/PayPalError.kt:21]: These codes are not accessible to merchants - we return the error as a `CoreSDKError` with a description, but the code will just be an Int to the merchant. I think this is the same in iOS. Does this make sense? In BT we just return a description in `BraintreeException`. Do we need these codes? Moving them all to the CoreSDKError would make them visible to the merchant, but wouldn't support modularity.

File: PayPalCheckout/src/main/res/color/paypal_dark_blue.xml [hunk 6]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (sarahkoop) [inline PayPalCheckout/src/main/res/color/paypal_dark_blue.xml:1]: We were missing some of the Button customization classes (they had been included from Native Checkout) so adding them now that Native Checkout was removed.

File: Core/src/main/java/com/paypal/android/core/CoreConfig.kt [hunk 7]
```diff
@@ -1,5 +1,13 @@
 package com.paypal.android.core
 
+/**
+ * Configuration for SDK requests.
+ *
+ * @property clientId your client ID from the PayPal Developer Portal
+ * @property clientSecret
```
Comment (cgdibble) [inline Core/src/main/java/com/paypal/android/core/CoreConfig.kt:7]: Does this need a description akin to `clientId` above?

**Review by cgdibble — APPROVED** [review]

File: Core/src/main/java/com/paypal/android/core/CoreConfig.kt (hunk 7 above)
Comment (jcnoriega) [reply Core/src/main/java/com/paypal/android/core/CoreConfig.kt:7]: there is a TODO in this class, that consist on deleting the `clientSecret`. We added it at the beginning as a temporary fix por authenticating api calls until `ordersV2` didnt need it anymore. I think they have fix it, so we might have to write a ticket for this.

File: PayPalCheckout/src/main/java/com/paypal/android/checkout/PayPalClient.kt (hunk 4 above)
Comment (sshropshire) [reply PayPalCheckout/src/main/java/com/paypal/android/checkout/PayPalClient.kt:45]: I'm not sure but we should be able to fully encapsulate this and all they'll have to do is set a listener right after instantiation. We can re-visit this once it's implemented and we can use BT as a guide.

File: PayPalCheckout/src/main/java/com/paypal/android/checkout/PayPalLifeCycleObserver.kt [hunk 8]
```diff
@@ -0,0 +1,12 @@
+package com.paypal.android.checkout
+
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleObserver
+import androidx.lifecycle.OnLifecycleEvent
+
+class PayPalLifeCycleObserver(private val payPalClient: PayPalClient) : LifecycleObserver {
+    @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
+    fun resume() {
+        payPalClient.handleBrowserSwitchResult()
```
Comment (sshropshire) [inline PayPalCheckout/src/main/java/com/paypal/android/checkout/PayPalLifeCycleObserver.kt:10]: I haven't tested this but apparently, the lifecycle gets notified after `onCreate()` returns so if we can get merchants to call `setListener()` immediately after instantiation within onCreate we should be good.

File: PayPalCheckout/src/main/java/com/paypal/android/checkout/error/PayPalError.kt (hunk 5 above)
Comment (sshropshire) [reply PayPalCheckout/src/main/java/com/paypal/android/checkout/error/PayPalError.kt:21]: There's also this if we need it [Library Annotation](https://developer.android.com/studio/write/annotations#libraries) 🙌

File: build.gradle [hunk 9]
```diff
@@ -61,9 +61,10 @@ buildscript {
             "jsonAssert"                  : "org.skyscreamer:jsonassert:1.4.0",
             "striktMockk"                 : "io.strikt:strikt-mockk:0.30.1",
             "striktCore"                  : "io.strikt:strikt-core:0.30.1",
+            "jsonMock"                    : "org.json:json:20211205",
 
-            // PayPal
-            "paypalNativeCheckout"        : "com.paypal.checkout:android-sdk:0.5.2"
+            // Browser Switch
+            "browserSwitch"               : "com.braintreepayments.api:browser-switch:2.1.1"
```
Comment (sshropshire) [inline build.gradle]: nit: can we move this up above the test dependencies?

**Review by sshropshire — APPROVED** [review]

**Comment by nyteshade** [comment]
Please do not merge these until I have spoken to Aki and David

File: PayPalCheckout/src/main/java/com/paypal/android/checkout/PayPalLifeCycleObserver.kt (hunk 8 above)
Comment (sarahkoop) [reply PayPalCheckout/src/main/java/com/paypal/android/checkout/PayPalLifeCycleObserver.kt:10]: But do we want to have that requirement on them? Would it be better to allow them to set the listener whenever?

File: build.gradle (hunk 9 above)
Comment (sarahkoop) [reply build.gradle]: done 👍

**Pull request closed without merging**